File: mwc/romana/relic/d/lib/libm/pow.c
/*
 * libm/pow.c
 * Raise x to the power y.
 */

#include <math.h>

#if	EMU87
#include "emumath.h"
#endif

double
pow(x, y)
double x, y;
{
	double r;
	register unsigned s, i, e;

	s = 0;
	i = 0;
	if (x == 0.0) {
		if (y <= 0.0)
			errno = EDOM;
		return 0.0;
	}
	if ((r = modf(y, &r)) < 0.0)
		r += 1.0;
	if (x < 0.0) {
		if (r != 0.0) {
			errno = EDOM;
			return 0.0;
		}
		x = -x;
		if (((int) y) & 1)
			s = 1;
	}
	if (y < 0.0) {
		y = -y;
		i = 1;
	}
	if (r!=0.0 || y>16384.0)
		r = _two(y*log10(x)*LOG10B2);
	else {
		r = 1.0;
		for (e=y; e; e>>=1) {
			if (e&01)
				r *= x;
			x *= x;
		}
	}
	if (i)
		r = 1/r;
	return s ? -r : r;
}

/* end of libm/pow.c */

File: doc/mwc/doc/coherent/lx/pow.m
.TH pow() "" "" "Mathematics Function (libm)"
.PC "Compute a power of a number"
.B "#include <math.h>"
\fBdouble pow(\fIz\^\fB, \fIx\^\fB)\fR
\fBdouble \fIz\^\fB, \fIx\^\fB;\fR
.PP
.B pow()
returns \fIz\fR raised to the power of \fIx\fR, or \fIz\fB^\fRx\fR.
If an overflow error occurs (that is, you attempt to compute a number
that is too large to fit into a double-precision floating-point number),
.B pow()
returns a huge value and sets
.B errno
to
.BR ERANGE .
.SH Example
For an example of this function, see the entry for
.BR log10() .
.SH "See Also"
.Xr "libm" libm
.br
\*(AS, \(sc4.5.5.1
.br
\*(PX Standard, \(sc8.1
